Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -214,71 +361,114 @@ static RefPtr<SharedBuffer> parseResponseMessage(const String& message)
+        opencdm_session_close(session);
+        opencdm_destruct_session(session);
+    });
+    m_id = String::fromUTF8(opencdm_session_id(m_session.get()));
+    callbacksNotifier.addObserver(m_session.get(), this);
+}
 
-    m_isValid = temporaryURLLength;
+CDMInstanceOpenCDM::Session::~Session()
+{
+    OCDMSessionCallbacksNotifier::singleton().removeObserver(this);
+}
 
+void CDMInstanceOpenCDM::Session::didGenerateChallengeCallback(String&& message)
+{
     // We could do all operations with String but this way we can save some copies.
     size_t typePosition = message.find(":Type:");
-    String requestType(message.c_str(), typePosition != std::string::npos ? typePosition : 0);
+    String requestType(message.characters8(), typePosition != notFound ? typePosition : 0);
     unsigned offset = 0;
-    if (!requestType.isEmpty() && requestType.length() != message.size())
+    if (!requestType.isEmpty() && requestType.length() != message.length())
         offset = typePosition + 6;
 
-    m_message = SharedBuffer::create(message.c_str() + offset, message.size() - offset);
+    m_message = SharedBuffer::create(message.characters8() + offset, message.length() - offset);
     m_needsIndividualization = requestType.length() == 1 && static_cast<WebCore::MediaKeyMessageType>(requestType.toInt()) == CDMInstance::MessageType::IndividualizationRequest;
+    for (const auto& challengeCallback : m_challengeCallbacks)
+        challengeCallback(this);
+    m_challengeCallbacks.clear();
+}
+
+void CDMInstanceOpenCDM::Session::didUpdateKeyCallback(String&& status)
+{
+    for (auto& keyStatusData : m_keyStatusDatas) {
+        m_lastStatus = opencdm_session_status(m_session.get(), keyStatusData.m_key.get(), keyStatusData.m_keySize);
+        keyStatusData.m_callback(this, m_lastStatus, status);
+    }
+    m_keyStatusDatas.clear();
+
+    for (auto& loadCallback : m_loadCallbacks)
+        loadCallback(this, true, status);
+    m_loadCallbacks.clear();
+
+    for (auto& removeCallback : m_removeCallbacks)
+        removeCallback(this, true, status);
+    m_removeCallbacks.clear();
+}
+
+void CDMInstanceOpenCDM::Session::didReceiveMessageCallback(String&& message)
+{
+}
+
+void CDMInstanceOpenCDM::Session::generateChallenge(DidGenerateChallengeCallback&& callback)
+{
+    if (isValid()) {
+        callback(this);
+        return;
+    }
+
+    m_challengeCallbacks.append(WTFMove(callback));
 }
 
-std::pair<media::OpenCdm::KeyStatus, String> CDMInstanceOpenCDM::Session::update(const uint8_t* data, const unsigned length)
+void CDMInstanceOpenCDM::Session::update(const uint8_t* data, const unsigned length, DidUpdateKeyCallback&& callback)
 {
-    std::string response;
-    m_lastStatus = m_session.Update(data, length, response);
-    return std::make_pair(m_lastStatus, String(response.c_str(), response.empty() ? 0 : response.size()));
+    KeyUpdateData updateData(WTFMove(callback), data, length);
+    m_keyStatusDatas.append(WTFMove(updateData));
```

Comment:
Maybe you can avoid the explicit creation of the `KeyUpdateData` and move it later by creating it either with {} or by invoking the constructor in the append call.